STM32G4x/STM32L4Rx/STM32L4Sx: Fix sys clock frequency and boost voltage for highest clock frequencies

FRASTM opened the pull request:
This patch enables the main regulator boost voltage for sys clock highest frequency.

For STM32G4xx series, boost mode when sys clock is higher than 150MHz
For STM32L4Rx or STM32L4Sx series, boost mode when sys clock is higher than 80MHz

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/22245

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -305,6 +305,13 @@ static int stm32_clock_control_init(struct device *dev)
 	LL_UTILS_PLLInitTypeDef s_PLLInitStruct;
 
+#ifdef CONFIG_SOC_SERIES_STM32G4X
+	/* set power boost mode */
```
Let's move this at the very end of the function.

FRASTM replied:
This voltage regulator must be set to enable range1 Boost mode **before** enabling the PLL.
So in any case, before the `LL_PLL_ConfigSystemClock_HSI()` function in the stm32_clock_control_init

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -53,6 +53,12 @@ static inline void uart_stm32_set_baudrate(struct device *dev, u32_t baud_rate)
 
+#if defined(CONFIG_LPUART_1) \
+	&& defined(CONFIG_SOC_SERIES_STM32G4X) \
+	&& defined(CONFIG_UART_CONSOLE)
```
I wonder why the dependency on `CONFIG_UART_CONSOLE`.

FRASTM replied:
this fix will output correctly on console on nucleo_g431rb (where console is assigned to LPUART1). Other use cases are not considered (tested).

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -665,6 +671,13 @@ static int uart_stm32_init(struct device *dev)
+	&& defined(CONFIG_SOC_SERIES_STM32G4X) \
+	&& defined(CONFIG_UART_CONSOLE)
+	/* the clock source for LPUART console is set to HSI */
+	LL_APB1_GRP2_EnableClock(LL_APB1_GRP2_PERIPH_LPUART1);
```
Should be done by stm32_clock_control_on.

FRASTM replied:
already done in stm32_clock_control_on, so removed there

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -665,6 +671,13 @@ static int uart_stm32_init(struct device *dev)
+	&& defined(CONFIG_UART_CONSOLE)
+	/* the clock source for LPUART console is set to HSI */
+	LL_APB1_GRP2_EnableClock(LL_APB1_GRP2_PERIPH_LPUART1);
+	LL_RCC_SetLPUARTClockSource(LL_RCC_LPUART1_CLKSOURCE_HSI);
```
I'm a bit concerned: by doing it that way, we're introducing hard coded configuration which doesn't match default SoC configuration.
We should be able to use default source clck configuration, which doesn't require all this.
Can you elaborate why default PCLK source for LPUART doesn't work ?

FRASTM replied:
requires register access to g431 to investigate further

FRASTM commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -305,6 +305,13 @@ static int stm32_clock_control_init(struct device *dev)
 
+#ifdef CONFIG_SOC_SERIES_STM32G4X
+	/* set power boost mode */
+	if (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC >= 170000000) {
```
ok

FRASTM commented on soc/arm/st_stm32/stm32g4/soc.c:
```diff
@@ -41,6 +41,10 @@ static int stm32g4_init(struct device *arg)
 	SystemCoreClock = 16000000;
 
+#ifdef CONFIG_DEBUG
+	LL_DBGMCU_EnableDBGSleepMode();
```
ok

FRASTM commented:
power boost does not impact clock control but only valid for stm32g4.

rosterloh commented:
Not sure if this is helpful to this issue but I needed to change stm32g4xx_ll_utils.c in the stm32 hal in order to get the correct baud rate. 
`
status = UTILS_EnablePLLAndSwitchSystem(pllfreq, UTILS_ClkInitStruct);
`
Needs to be added to line 335. After AHBCLK divider is set back to 1 the system clock variable is not updated so while it's running at the correct speed the functions the request the clock value are returning half the actual value.

rosterloh commented on drivers/clock_control/clock_stm32g4.c:
```diff
@@ -29,6 +29,11 @@ void config_pll_init(LL_UTILS_PLLInitTypeDef *pllinit)
 	pllinit->PLLR = pllr(CONFIG_CLOCK_STM32_PLL_R_DIVISOR);
+
+	/* set power boost mode */
+	if (sys_clock_hw_cycles_per_sec() >= 170000000) {
```
Shouldn't this be 150000000?

FRASTM replied:
yes, 150MHz

martinjaeger commented:
So this PR attempts to fix also #21681, which was not mentioned yet.

I agree with @rosterloh that there is a bug in the STM LL driver, see also [his PR](https://github.com/zephyrproject-rtos/hal_stm32/pull/36). I think that's the root cause of the wrong UART baud rate. As far as I could see, the MCU actually runs at the specified speed, but only SystemCoreClock variable is set wrong in LL drivers. Did quite a lot of debugging looking at the actual register contents.

FRASTM commented:
stm32g4xx: set power boost mode for sys clock greater than 150MHz

FRASTM commented:
updating west.yml after merge of the https://github.com/zephyrproject-rtos/hal_stm32/pull/37

FRASTM committed:
west.yml: update hal stm32g4xx and stm32l4R/stm32l4S modules

This updates the stm32cube drivers to fix the issue on sysclock

Signed-off-by: Francois Ramu <francois.ramu@st.com>

MaureenHelm reviewed (changes requested):
Please update the PR summary to include which issues it fixes. i.e., "Fixes #NNNN"

FRASTM committed:
driver: clock: stm32: boost voltage when clock exceeds 150MHz

This patch activates the boost mode for the main regulator
when the system frequency above 150MHz for stm32g4xx soc series.
To save power, the boost mode should be disabled below 150MHz.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
driver: clock: stm32: boost voltage when clock exceeds 80MHz

This patch activates the boost mode for the main regulator
when the system frequency above 80MHz for stm32l4Rx/stm32l4Sx
soc series.
To save power, the boost mode should be disabled below 80MHz.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on drivers/clock_control/clock_stm32l4_wb.c:
```diff
@@ -30,6 +30,12 @@ void config_pll_init(LL_UTILS_PLLInitTypeDef *pllinit)
 	pllinit->PLLR = pllr(CONFIG_CLOCK_STM32_PLL_R_DIVISOR);
+#ifdef PWR_CR5_R1MODE
+	/* set power boost mode for sys clock greater than 80MHz */
+	if (sys_clock_hw_cycles_per_sec() >= 80000000) {
```
changed to MHZ(80)

FRASTM commented on drivers/clock_control/clock_stm32g4.c:
```diff
@@ -29,6 +29,11 @@ void config_pll_init(LL_UTILS_PLLInitTypeDef *pllinit)
 	pllinit->PLLR = pllr(CONFIG_CLOCK_STM32_PLL_R_DIVISOR);
+
+	/* set power boost mode for sys clock greater than 150MHz */
+	if (sys_clock_hw_cycles_per_sec() >= 150000000) {
```
changed to MHZ(150)

FRASTM commented:
change to MHZ(xxx)
and describe PR : Fixes https://github.com/zephyrproject-rtos/zephyr/issues/22245
